src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionProvider.cs

```diff
@@ -47,51 +51,169 @@ public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeAct
             return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
         }
 
+        var (startElementNode, endElementNode) = GetStartAndEndElements(context, syntaxTree, _logger);
+
+        // Make sure the selection starts on an element tag
+        if (startElementNode is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (!TryGetNamespace(context.CodeDocument, out var @namespace))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var actionParams = CreateInitialActionParams(context, startElementNode, @namespace);
+
+        if (IsMultiPointSelection(context.Request.Range))
+        {
+            ProcessMultiPointSelection(startElementNode, endElementNode, actionParams);
+        }
+
+        var resolutionParams = new RazorCodeActionResolutionParams()
+        {
+            Action = LanguageServerConstants.CodeActions.ExtractToNewComponentAction,
+            Language = LanguageServerConstants.CodeActions.Languages.Razor,
+            Data = actionParams,
+        };
+
+        var codeAction = RazorCodeActionFactory.CreateExtractToNewComponent(resolutionParams);
+        return Task.FromResult<ImmutableArray<RazorVSInternalCodeAction>>([codeAction]);
+    }
+
+    private static (MarkupElementSyntax? Start, MarkupElementSyntax? End) GetStartAndEndElements(RazorCodeActionContext context, RazorSyntaxTree syntaxTree, ILogger logger)
+    {
         var owner = syntaxTree.Root.FindInnermostNode(context.Location.AbsoluteIndex, includeWhitespace: true);
         if (owner is null)
         {
-            _logger.LogWarning($"Owner should never be null.");
-            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+            logger.LogWarning($"Owner should never be null.");
+            return (null, null);
+        }
+
+        var startElementNode = owner.FirstAncestorOrSelf<MarkupElementSyntax>();
+        if (startElementNode is null || IsInsideProperHtmlContent(context, startElementNode))
+        {
+            return (null, null);
         }
 
-        var componentNode = owner.FirstAncestorOrSelf<MarkupElementSyntax>();
+        var endElementNode = GetEndElementNode(context, syntaxTree);
+        return (startElementNode, endElementNode);
+    }
 
-        // Make sure we've found tag
-        if (componentNode is null)
+    private static bool IsInsideProperHtmlContent(RazorCodeActionContext context, MarkupElementSyntax startElementNode)
+    {
+        // If the provider executes before the user/completion inserts an end tag, the below return fails
+        if (startElementNode.EndTag.IsMissing)
```

Comment:
Does this mean we can't handle self-closing tags, like `<Comp$$onent />`?

Reply:
Nope, this check corrects for a case where the provider executes while in the middle of typing an end tag, or right after inserting a starting tag (in those few seconds where the completion doesn't insert its corresponding end tag yet).

Anyway, this is removed in the latest PR where instead of doing this check I check for error diagnostics on a `markupSyntaxNode`.